libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2132,6 +2135,11 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 			ssl.append("Function is declared here:", functionPointerType->declaration().location());
 			if (functionPointerType->declaration().scope() == m_currentContract)
 				msg += " Did you forget to prefix \"this.\"?";
+			else if (contains(
+				m_currentContract->annotation().linearizedBaseContracts,
+				functionPointerType->declaration().scope()
+			))
+				msg += " Functions from base contracts have to be external.";
```

Comment:
Ok, this is actually true. But if it works if you assign it to an external function pointer, then it should also work without, shouldn't it? Maybe we can perform an implicit conversion or something.

Reply:
Yeah, that would be even better. I thought about suggesting that but I wasn't sure how complicated it would be.

But I just checked to make sure this workaround would even work - and turns out it does not. We don't have a contract address on a function pointer referenced like that so you can't directly assign it to a function pointer:
```solidity
contract Base {
    function baseFunctionPublic(uint) public pure {}
}

contract C is Base {
    function viaBaseDeclaration() public pure  {
        function(uint) external pure fPtr = Base.baseFunctionPublic;
        abi.encodeCall(fPtr, (1));
    }
}
```
```
Error: Type function (uint256) pure is not implicitly convertible to expected type function (uint256) pure external. Special functions can not be converted to function types.
 --> test.sol:7:9:
  |
7 |         function(uint) external pure fPtr = Base.baseFunctionPublic;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^```

Reply:
Why are base contract public functions not supported by `abi.encodeCall`?